pkg/controllers/cassandra/pilot/pilot.go

```diff
@@ -0,0 +1,182 @@
+package pilot
+
+import (
+	"fmt"
+	"hash/fnv"
+	"reflect"
+
+	"github.com/jetstack/navigator/pkg/apis/navigator/v1alpha1"
+	navigator "github.com/jetstack/navigator/pkg/client/clientset/versioned"
+	navlisters "github.com/jetstack/navigator/pkg/client/listers/navigator/v1alpha1"
+	"github.com/jetstack/navigator/pkg/controllers"
+	"github.com/jetstack/navigator/pkg/controllers/cassandra/util"
+	hashutil "github.com/jetstack/navigator/pkg/util/hash"
+	"k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	appslisters "k8s.io/client-go/listers/apps/v1beta1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/record"
+)
+
+const (
+	HashAnnotationKey = "navigator.jetstack.io/cassandra-pilot-hash"
+)
+
+type Interface interface {
+	Sync(*v1alpha1.CassandraCluster) error
+}
+
+type pilotControl struct {
+	naviClient   navigator.Interface
+	pilots       navlisters.PilotLister
+	pods         corelisters.PodLister
+	statefulSets appslisters.StatefulSetLister
+	recorder     record.EventRecorder
+}
+
+var _ Interface = &pilotControl{}
+
+func NewControl(
+	naviClient navigator.Interface,
+	pilots navlisters.PilotLister,
+	pods corelisters.PodLister,
+	statefulSets appslisters.StatefulSetLister,
+	recorder record.EventRecorder,
+) *pilotControl {
+	return &pilotControl{
+		naviClient:   naviClient,
+		pilots:       pilots,
+		pods:         pods,
+		statefulSets: statefulSets,
+		recorder:     recorder,
+	}
+
+}
+
+func (c *pilotControl) clusterPods(cluster *v1alpha1.CassandraCluster) ([]*v1.Pod, error) {
+	var clusterPods []*v1.Pod
+	allPods, err := c.pods.Pods(cluster.Namespace).List(labels.Everything())
+	if err != nil {
+		return clusterPods, err
+	}
+	for _, pod := range allPods {
+		podControlledByCluster, err := controllers.PodControlledByCluster(
+			cluster,
+			pod,
+			c.statefulSets,
+		)
+		if err != nil {
+			return clusterPods, err
+		}
+		if !podControlledByCluster {
+			continue
+		}
+		clusterPods = append(clusterPods, pod)
+	}
+	return clusterPods, nil
+}
+
+func (c *pilotControl) createOrUpdatePilot(cluster *v1alpha1.CassandraCluster, pod *v1.Pod) error {
+	desiredPilot := PilotForCluster(cluster, pod)
+	client := c.naviClient.NavigatorV1alpha1().Pilots(desiredPilot.GetNamespace())
+	lister := c.pilots.Pilots(desiredPilot.GetNamespace())
+	existingPilot, err := lister.Get(desiredPilot.GetName())
+	if k8sErrors.IsNotFound(err) {
+		_, err = client.Create(desiredPilot)
+		return err
+	}
+	if err != nil {
+		return err
+	}
+	err = util.OwnerCheck(existingPilot, cluster)
+	if err != nil {
+		return err
```

Comment:
TODO: should there be some way to detect that this function has failed because the existing pilot is owned by another cluster? (not important for merge)